File: src/shared/media/audio-mix.ts
/**
 * Mixing audio the suite made into a clip's own sound.
 *
 * The one place the pipeline's copy-only rule bends, and only on request
 * (`audio-plan.ts`). The clip's AAC packets that overlap the export window are
 * decoded with WebCodecs' `AudioDecoder` — straight from the samples the demux
 * already holds, so nothing re-reads the file — laid out at their own
 * timestamps, the bed is summed in, and the result is encoded once.
 *
 * Sync is kept at PARITY with the copy it replaces: decoded audio is placed at
 * the same packet times `copyAudio` would have written, so a mixed clip's own
 * sound sits exactly where a copied one would. The only new offset is this
 * encoder's own priming, which is taken off the front of the mix the same way
 * a bed is rendered ahead of it (`aacPrimingSeconds`).
 */

import type { Sample, Track } from 'mp4box';
import { aacPrimingSeconds } from './audio-encode';

/**
 * Planar floating-point audio — what `AudioBuffer` already is, and all an
 * encoder needs to read. Decoded and mixed audio are built as this shape
 * rather than as `AudioBuffer`s, which cannot be constructed off the main
 * thread in every browser.
 */
export interface PlanarAudio {
  sampleRate: number;
  numberOfChannels: number;
  length: number;
  getChannelData(channel: number): Float32Array;
}

export function planar(sampleRate: number, channels: Float32Array[]): PlanarAudio {
  return {
    sampleRate,
    numberOfChannels: channels.length,
    length: channels[0]?.length ?? 0,
    getChannelData: (c) => channels[c],
  };
}

/**
 * The bed summed into the clip's sound, sample by sample, clamped to full
 * scale. Both must share a sample rate (the bed is rendered AT the clip's).
 *
 * Channels meet sensibly rather than by index alone: a mono clip takes the
 * average of a stereo bed; a stereo clip takes the bed's left and right on its
 * own; extra clip channels are left as they are. The output is the clip's
 * length and layout — the bed never makes a clip longer or wider.
 */
export function mixPlanar(clip: PlanarAudio, bed: PlanarAudio, bedGain = 1): PlanarAudio {
  const out: Float32Array[] = [];
  for (let c = 0; c < clip.numberOfChannels; c++) {
    const channel = new Float32Array(clip.getChannelData(c));
    let add: ((i: number) => number) | null = null;
    if (bed.numberOfChannels > 0) {
      if (clip.numberOfChannels === 1 && bed.numberOfChannels >= 2) {
        const l = bed.getChannelData(0);
        const r = bed.getChannelData(1);
        add = (i) => (l[i] + r[i]) / 2;
      } else if (c < bed.numberOfChannels) {
        const b = bed.getChannelData(c);
        add = (i) => b[i];
      } else if (bed.numberOfChannels === 1 && c < 2) {
        const b = bed.getChannelData(0);
        add = (i) => b[i];
      }
    }
    if (add) {
      const n = Math.min(channel.length, bed.length);
      for (let i = 0; i < n; i++) {
        const v = channel[i] + add(i) * bedGain;
        channel[i] = v > 1 ? 1 : v < -1 ? -1 : v;
      }
    }
    out.push(channel);
  }
  return planar(clip.sampleRate, out);
}

/** The clip's real AudioSpecificConfig from its sample entry, when mp4box exposes it. */
function aacDescription(sample: Sample): Uint8Array | undefined {
  const entry = sample.description as unknown as {
    esds?: { esd?: { descs?: { descs?: { data?: Uint8Array }[] }[] } };
  };
  const data = entry.esds?.esd?.descs?.[0]?.descs?.[0]?.data;
  return data && data.length >= 2 ? new Uint8Array(data) : undefined;
}

export type DecodeResult = { ok: true; audio: PlanarAudio } | { ok: false; reason: string };

const CANNOT_MIX =
  'The clip’s sound could not be mixed here, so it kept its own sound untouched and the ticks were left out.';

/**
 * Decode `[baseMicros, endMicros)` of the clip's AAC track into planar audio,
 * `endMicros - baseMicros` long, shifted EARLY by this encoder's priming so the
 * re-encoded mix lands where the copy would have.
 */
export async function decodeAacWindow(
  track: Track,
  samples: readonly Sample[],
  baseMicros: number,
  endMicros: number,
  fallbackDescription: (sampleRate: number, channels: number) => Uint8Array,
): Promise<DecodeResult> {
  const sampleRate = track.audio?.sample_rate ?? 48000;
  const numberOfChannels = track.audio?.channel_count ?? 2;
  if (typeof AudioDecoder === 'undefined' || typeof AudioData === 'undefined') {
    return { ok: false, reason: CANNOT_MIX };
  }
  const first = samples.find((s) => s.data);
  if (!first) return { ok: false, reason: CANNOT_MIX };

  const config: AudioDecoderConfig = {
    codec: 'mp4a.40.2',
    sampleRate,
    numberOfChannels,
    description: aacDescription(first) ?? fallbackDescription(sampleRate, numberOfChannels),
  };
  const probe = await AudioDecoder.isConfigSupported(config).catch(() => null);
  if (!probe?.supported) return { ok: false, reason: CANNOT_MIX };

  const lead = aacPrimingSeconds(sampleRate) * 1_000_000;
  const origin = baseMicros + lead;
  const length = Math.max(1, Math.ceil(((endMicros - baseMicros) / 1_000_000) * sampleRate));
  const channels = Array.from({ length: numberOfChannels }, () => new Float32Array(length));

  let failure: unknown = null;
  const decoder = new AudioDecoder({
    output: (data) => {
      try {
        const frames = data.numberOfFrames;
        const at = Math.round(((data.timestamp - origin) / 1_000_000) * sampleRate);
        const layout = Math.min(numberOfChannels, data.numberOfChannels);
        for (let c = 0; c < layout; c++) {
          const plane = new Float32Array(frames);
          data.copyTo(plane, { planeIndex: c, format: 'f32-planar' });
          const from = Math.max(0, -at);
          const to = Math.min(frames, length - at);
          if (to > from) channels[c].set(plane.subarray(from, to), at + from);
        }
      } finally {
        data.close();
      }
    },
    error: (e) => {
      failure = e;
    },
  });

  try {
    decoder.configure(config);
    const toMicros = (v: number, timescale: number) => Math.round((v / timescale) * 1_000_000);
    // One packet of pre-roll before the window: an AAC frame decodes against
    // its predecessor, and the first one after a cold start is not clean.
    const preroll = 1;
    let started = -1;
    samples.forEach((sample, i) => {
      if (!sample.data) return;
      const cts = toMicros(sample.cts, sample.timescale);
      const duration = toMicros(sample.duration, sample.timescale);
      const inWindow = cts + duration > origin && cts < endMicros + lead;
      if (!inWindow) return;
      if (started < 0) {
        started = i;
        for (let p = Math.max(0, i - preroll); p < i; p++) {
          const pre = samples[p];
          if (!pre.data) continue;
          decoder.decode(
            new EncodedAudioChunk({
              type: 'key',
              timestamp: toMicros(pre.cts, pre.timescale),
              duration: toMicros(pre.duration, pre.timescale),
              data: pre.data,
            }),
          );
        }
      }
      decoder.decode(new EncodedAudioChunk({ type: 'key', timestamp: cts, duration, data: sample.data }));
    });
    await decoder.flush();
  } catch (e) {
    failure = failure ?? e;
  } finally {
    if (decoder.state !== 'closed') decoder.close();
  }

  if (failure) return { ok: false, reason: CANNOT_MIX };
  return { ok: true, audio: planar(sampleRate, channels) };
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Media/AudioMix.swift
// Mixing audio the suite MADE into a clip's own sound — the pure half of
// `src/shared/media/audio-mix.ts`.
//
// The one place the pipeline's copy-only rule bends, and only on request
// (`AudioPlan.swift`'s `mix`). The web decodes the clip's AAC with
// WebCodecs (`decodeAacWindow`); the app decodes it with `AVAssetReader`
// (`apple/Atelier/Video/VideoExport.swift`), and both sum the bed in HERE,
// chunk by chunk or whole, so the channels meet by the same rules.

import Foundation

/// Planar floating-point audio — one array per channel, full scale ±1 — the
/// web's `PlanarAudio` (what `AudioBuffer` already is, and all an encoder
/// needs to read). Its length is its FIRST channel's, as the web's `planar`.
public struct PlanarAudio: Equatable, Sendable {
    public var sampleRate: Double
    public var channels: [[Float]]

    public init(sampleRate: Double, channels: [[Float]]) {
        self.sampleRate = sampleRate
        self.channels = channels
    }

    public var numberOfChannels: Int { channels.count }
    public var length: Int { channels.first?.count ?? 0 }
    /// Seconds at `sampleRate`; 0 when the rate is not a rate.
    public var seconds: Double { sampleRate > 0 ? Double(length) / sampleRate : 0 }

    /// The web's `getChannelData`.
    public func channelData(_ channel: Int) -> [Float] { channels[channel] }
}

/// The bed summed into the clip's sound, sample by sample, clamped to full
/// scale. Both must share a sample rate (the bed is rendered AT the clip's).
///
/// Channels meet sensibly rather than by index alone: a mono clip takes the
/// average of a stereo bed; a stereo clip takes the bed's left and right on
/// its own; a mono bed reaches the first two channels of a wider clip; extra
/// clip channels are left as they are. The output is the clip's length and
/// layout — the bed never makes a clip longer or wider — and the clip handed
/// in is never written to (a value type here, a copy there).
///
/// The sum is taken in doubles and stored as a Float, as the web sums a
/// `Float32Array`'s values in JavaScript numbers.
public func mixPlanar(_ clip: PlanarAudio, _ bed: PlanarAudio, _ bedGain: Double = 1) -> PlanarAudio {
    var out: [[Float]] = []
    out.reserveCapacity(clip.numberOfChannels)
    for c in 0..<clip.numberOfChannels {
        var channel = clip.channels[c]
        let add = bedSource(clipChannels: clip.numberOfChannels, channel: c, bed: bed)
        if let add {
            let n = min(channel.count, bed.length)
            for i in 0..<n {
                let v = Double(channel[i]) + add(i) * bedGain
                channel[i] = Float(v > 1 ? 1 : (v < -1 ? -1 : v))
            }
        }
        out.append(channel)
    }
    return PlanarAudio(sampleRate: clip.sampleRate, channels: out)
}

/// Which of the bed's samples feed clip channel `c`, or nil for none.
private func bedSource(clipChannels: Int, channel c: Int, bed: PlanarAudio) -> ((Int) -> Double)? {
    let bedChannels = bed.numberOfChannels
    if bedChannels == 0 { return nil }
    if clipChannels == 1 && bedChannels >= 2 {
        let l = bed.channels[0]
        let r = bed.channels[1]
        return { i in
            let left = i < l.count ? Double(l[i]) : 0
            let right = i < r.count ? Double(r[i]) : 0
            return (left + right) / 2
        }
    }
    if c < bedChannels {
        let b = bed.channels[c]
        return { i in i < b.count ? Double(b[i]) : 0 }
    }
    if bedChannels == 1 && c < 2 {
        let b = bed.channels[0]
        return { i in i < b.count ? Double(b[i]) : 0 }
    }
    return nil
}
